86cuerykj/auth

abhijith-biju opened the pull request:
## Overview
This pull request introduces addition authentication functionality for the application.

## Changes Made

### 1. authSlice
- Added authSlice to Redux store to manage user data and manage authentication modal state globally.
- Added actions for sign-in and sign-out.
- Implemented functionality to keep user signed-in on page reload.

### 2. Sign-in and Sign-up functionality
- Integrated the sign-in and sign-up apis.
- Handled error responses.
- Updated navbar component to be in sync with auth state.

### 3. Protected Routes
- Added role based access control for protecting customer and admin-only pages.

abhijith-biju committed:
feat(authState): add auth slice and auth modal
- setup auth state in store
- made auth modal state available globallu in app

abhijith-biju committed:
feat(signIn): add sign-in functionality
- updated login component to use formik-mui bindings
- add reducer functions to setCredentials and logout
- updated header component to be synchronized to auth state

abhijith-biju committed:
fix(signIn): bug when signIn is successful
- fix bug where auth modal not closing after logging in.

abhijith-biju committed:
merge 86cumak5j/API-layer into 86cuerykj/Auth
- changes to response interceptor
- removed unused store folder

abhijith-biju committed:
feat(signIn): implemented persistant login

abhijith-biju committed:
Merge branch 'develop' of github.com:amaljoeqb/bus-ticket-booking-react into 86cuerykj/Auth

abhijith-biju committed:
refactor(signIn): change return value of signIn endpoint
- changed firstName to fullName.

abhijith-biju committed:
feat(signIn): add error handling for sign-in
- add error handling when incorrect email or password is entered.

abhijith-biju committed:
refactor(signIn): move signIn submit handler to separate file
- fix bug when loading userData from local storage

abhijith-biju committed:
fix(signIn): fix bug for persistent login
- fix bug where user not logged out if expired token in storage.
- add error handling when decoding token.

amaljoeqb committed:
update: add ISignUpProps

abhijith-biju committed:
Merge branch '86cuerykj/Auth' of github.com:amaljoeqb/bus-ticket-booking-react into 86cuerykj/Auth

abhijith-biju committed:
feat(signUp): add signUp functionality
- integrated signUp api
- handled axios errors for signIn api

abhijith-biju committed:
feat(auth): add loaders and toasts
- add loaders for signIn and signUp
- add toasts for error and success responses

abhijith-biju committed:
fix(header): change profile style
- change profile menu style when logged in

abhijith-biju committed:
refactor(confirmDialog): updated confirmDialog component to use arrow function

abhijith-biju committed:
feat(logout): add confirmation modal for logout action

abhijith-biju committed:
feat(protectedRoutes): add requireAuth component
- add requireAuth component that redirects to home page when a protected route is accessed without being authenticated.

abhijith-biju committed:
refactor(protectedRoutes): add alert to auth modal
- replace error toast on unauthorized error with alert in signIn modal

abhijith-biju committed:
fix(signUp): change in password validation
- updated validation for password to match expected value from backend.

abhijith-biju committed:
fix(auth): bug where error toast not displayed
- fixed bug where error toast not displayed when unknown axios error occurs.
- added password requirements info alert in sign up modal.

abhijith-biju committed:
feat(protectedRoutes): add role based access control
- update RequireAuth component to include RABC.
- made booking page a protected route (admin, customer)
- made users page a admin-only route

abhijith-biju committed:
Merge branch 'develop' of github.com:amaljoeqb/bus-ticket-booking-react into 86cuerykj/Auth

abhijith-biju committed:
refactor(protectedRoute): changes to navigation on forbidden error
- navigate to home page with error toast instead of 404 page when

abhijith-biju committed:
refactor(auth): redirect to home page on logout
- remove length validation for password during sign-in.
- update forbidden error toast to support language change.

abhijith-biju committed:
chore(requireAuth): comments and changes to signIn
- added comment explaining the need for wrapping
forbidden error toast inside setTimeout.
- reset sign-in/sign-up on tab switch.

akshay-krishna commented on src/i18n/types/resources.ts:
```diff
@@ -13,6 +13,7 @@ const {
     filterSort,
     landingPage,
     auth,
+    logoutModal,
```
Better to change it to logoutConfirmationModal for better understanding.

abhijith-biju replied:
Updated.

akshay-krishna commented on src/components/ui/AuthModal/SignUp/validationSchema.ts:
```diff
@@ -22,6 +22,10 @@ const getValidationSchema = (t: TFunction<'auth'>) =>
         password: Yup.string()
             .trim()
+            .matches(
+                /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[~`!@#$%^&*()-_+={}[\]|\\;:'"<>,./?]).{8,}$/,
```
Wrong validation even when more than 6 characters are entered.

abhijith-biju replied:
Resolved.

akshay-krishna reviewed (commented):
Add changeLog

RJR-003 commented on src/layout/Header/Header.tsx:
```diff
@@ -1,57 +1,63 @@
+    ListItemIcon,
 } from '@mui/material';
+import { NavLink, useNavigate } from 'react-router-dom';
+import { StyledButton } from '../Footer/Footer.styled';
```
You could add a seperate styledButton for header instead of using from footer component. Or use a shared button.

abhijith-biju replied:
Updated.

RJR-003 commented on src/layout/Header/Header.tsx:
```diff
@@ -1,57 +1,63 @@
-            setLoginClick(false);
-            setIsLoginModalOpen(false);
-        }
+    const handleLogoutClick = () => {
```
Please provide a logout toast to notify the user.

abhijith-biju replied:
Added.

abhijith-biju commented:
> Add changeLog

updated changelog.

abhijith-biju committed:
fix(auth): resolve mr comments
- add toast on logout.
- fix bug in sign-up password validation.
- updated changelog.
